src/lint.cpp

```diff
@@ -500,11 +519,21 @@ void linter::propagate_variable_uses_to_parent_scope(
                var.name.normalized_name();
   };
 
+  bool has_eval = this->is_eval_used();
   for (const used_variable &used_var : current_scope.variables_used) {
     QLJS_ASSERT(!current_scope.declared_variables.find(used_var.name));
     const declared_variable *var =
         parent_scope.declared_variables.find(used_var.name);
-    if (var) {
+    if (used_var.name.normalized_name() == u8"eval") {
+      if (propagate_eval_use) {
+        (allow_variable_use_before_declaration
+             ? parent_scope.variables_used_in_descendant_scope
+             : parent_scope.variables_used)
+            .emplace_back(used_var);
+      }
+      break;
+    } else if (has_eval) {
+    } else if (var) {
```

Comment:
This is a bit odd. `this->is_eval_used` loops through `current_scope.variables_used` to find `"eval"`. Then this code loops through `current_scope.variables_used` again to find `"eval"`, skipping other variables if we found `"eval"` previously.

I think you meant to write this code differently, e.g. by skipping the second loop entirely if we find `"eval"` in the first loop:
```suggestion
bool has_eval = false;
for (const used_variable &used_var : current_scope.variables_used) {
      if (used_var.name.normalized_name() == u8"eval") {
          (allow_variable_use_before_declaration
             ? parent_scope.variables_used_in_descendant_scope
             : parent_scope.variables_used)
            .emplace_back(used_var);
          has_eval = true;
          break;
      }
}
if (!has_eval) {
  for (const used_variable &used_var : current_scope.variables_used) {
    QLJS_ASSERT(!current_scope.declared_variables.find(used_var.name));
    const declared_variable *var =
        parent_scope.declared_variables.find(used_var.name);
    if (var) {
```